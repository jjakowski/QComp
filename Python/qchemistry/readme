

#  slicing test: 
#  II =  np.reshape([i1+ i2*10 +i3*100 +i4*1000  for i4 in range(4) for i3 in range(4) for  i2 in range(4) for i1 in range(4)], (4,4,4,4) )
#
#  III =  np.reshape([i1+ i2*10 +i3*100 +i4*1000  for i4 in range(8) for i3 in range(8) for  i2 in range(8) for i1 in range(8)], (8,8,8,8) )
#
#    octave style slicing:
#   II  [[0,2],:,:,:]   [:,[1,3],:,:]   [:,:,[0,3],:]   [:,:,:,[1,2]]
# ----  
#  For 4 electrons (Nel=4_ generate  single excitaitons (Nexc=1)  from the occ_list = [1,2,3,4]  into virt_list=[5,6,7,8,9,10]
#
#  from itertools import combinations  
#   Nel = 4 ; Nexc=  1; CIS_list = [ list(x+y)   for x in  combinations(occ_list, Nel-Nexc )  for y in  combinations(virt_list, Nexc)]
#
#  ...and double excitations:
#
#   Nel = 4 ; Nexc=  2; CID_list =[ list(x+y)   for x in  combinations(occ_list, Nel-Nexc )  for y in  combinations(virt_list, Nexc)]
#
#  CISD_List = CIS_list + CID_list 
#---- 
# HOWTO reload function (debugging):
#    let assume that my function "get_CI_ham" (debugged)  is in   "tmp.py"
#  
#   #main.py" 
#    import imp 
#    import tmp 
#    tmp.get_CI_ham( ...)     # old version
#
#    imp.reload(tmp)
#    tmp.get_CI_ham( ...)     # new  version
#
#---  example 2:
#      import imp  
#      from Modules  import hello 
#      

>>> from Modules  import hello 
>>> import imp
>>> hello.world()
Hello, World!
>>> hello.myfun(1,2)     
(2, 6)
#
# here " vi ./Modules/hello.py"  change  myfun()  function
#
>>> imp.reload(hello)
<module 'Modules.hello' from '/Users/j2c/Dropbox (ORNL)/P3HT-transport/H2-test/Modules/hello.py'>
>>> hello.myfun(1,2)
(2, 4)
>>> 


#----
import psi4
import numpy as np

import myutils


#  adding point charges:
# https://github.com/psi4/psi4/blob/master/tests/extern1/input.dat

molecule water {
  0 1
  O  -0.778803000000  0.000000000000  1.132683000000
  H  -0.666682000000  0.764099000000  1.706291000000
  H  -0.666682000000  -0.764099000000  1.706290000000
  symmetry c1
  no_reorient
  no_com
}


H2a_geometry = psi4.geometry ("""
0 1
H   0.000000   0.0      0.0
H   0.8        0.0      0.0 
symmetry c1
no_reorient
no_com
""")



LiHa_geometry = psi4.geometry ("""
0 1
Li   0.000000   0.0      0.0
H   1.0        0.0      0.0 
symmetry c1
no_reorient
no_com
""")


LiHb_geometry = psi4.geometry ("""
0 1
Li   0.000000   0.0      0.0
H   0.0        1.0      0.0 
symmetry c1
no_reorient
no_com
""")


LiHc_geometry = psi4.geometry ("""
0 1
Li   0.000000   0.0      0.0
H   0.0        0.0      1.0 
symmetry c1
no_reorient
no_com
""")




#------------------------------------------------------------------------------------------
# Get MO coefficients from SCF wavefunction
#
#scf:
#basis = '6-311G', '6-31G(d)', 'sto-3g',

psi4.set_options({'basis':        'sto-3g',
                  'scf_type':     'pk',
                  'mp2_type':     'conv',
                  'e_convergence': 1e-8,
                  'd_convergence': 1e-8})

#  Solve SCF: # Get the SCF wavefunction & energies
scf_e, scf_wfn = psi4.energy('scf', return_wfn=True,molecule=H2_geometry)
print("E(SCF):", scf_e, scf_wfn)

#nuclear repulsion energy
E_nucl = H2_geometry.nuclear_repulsion_energy()
print("Nuclear repulsion:", E_nucl)

#- get matrices  in a.o.
dir(scf_wfn)
Ovelap = np.array(scf_wfn.S())
Fa     =  np.array(scf_wfn.Fa())
H      =  np.array(scf_wfn.H())

#------------------------------------------------
#   various integrals

# Create instance of MintsHelper class:
mints   = psi4.core.MintsHelper(scf_wfn.basisset())
mints_a = psi4.core.MintsHelper(scf_wfn_a.basisset())
mints_b = psi4.core.MintsHelper(scf_wfn_b.basisset())
mints_c = psi4.core.MintsHelper(scf_wfn_c.basisset())

# Build core Hamiltonian
T = np.asarray(mints.ao_kinetic())
V = np.asarray(mints.ao_potential())
H_core_ao = T + V

#- dipole moment matrix
mux,muy,muz  = mints.ao_dipole()
print(np.array(mux)) 
print(np.array(muy))
print(np.array(muz))
# ==> ERIs <==


 


# 
