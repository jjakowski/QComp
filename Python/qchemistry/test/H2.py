
#from  qcpy import  *
from qchemistry.qcpy import  *

import psi4
import numpy as np


# Initialize structure
H2_geometry = psi4.geometry ("""
0 1
H   0.000000   0.0      0.0
H   0.0        0.0      0.7474 
symmetry c1
""")

#basis = '6-311G'
#psi4.set_options({'basis':        'sto-3g-jacek',
psi4.set_options({'basis':        'sto-3g',
                  'scf_type':     'pk',
                  'mp2_type':     'conv',
                  'e_convergence': 1e-8,
                  'd_convergence': 1e-8})


#-----------------------------------------------
# Solve scf:
#-----------------------------------------------

# Get the SCF wavefunction & energies
scf_e, scf_wfn = psi4.energy('scf', return_wfn=True,molecule=H2_geometry)

print("E(SCF):", scf_e, scf_wfn)

print("")
print('Alpha orbital energies: ', np.array(scf_wfn.epsilon_a()))
print('Beta  orbital energies: ', np.array(scf_wfn.epsilon_b()))
#print(scf_wfn.doccpi() )

print("")
print("N-alpha-electrons:  %d "  % scf_wfn.nalpha())
print("N-beta--electrons:  %d "  % scf_wfn.nbeta())

print("")
print("Alpha occupations: ", np.asarray(scf_wfn.occupation_a()))
print("Beta  occupations: ", np.asarray(scf_wfn.occupation_b()))

#nuclear repulsion energy
E_nucl = H2_geometry.nuclear_repulsion_energy()
print("")
print("Nuclear repulsion:", E_nucl)



#
#-----------------------------------------------
#  Electron Repulsion integrals:¶
#-----------------------------------------------

# Get MO coefficients from SCF wavefunction

# ==> ERIs <==
# Create instance of MintsHelper class:
mints = psi4.core.MintsHelper(scf_wfn.basisset())


nbf = mints.nbf()           # number of basis functions
nso = 2 * nbf               # number of spin orbitals
nalpha = scf_wfn.nalpha()   # number of alpha electrons
nbeta = scf_wfn.nbeta()     # number of beta electrons
nocc = nalpha + nbeta       # number of occupied orbitals
nvirt = 2 * nbf - nocc      # number of virtual orbitals
list_occ_alpha= np.asarray(scf_wfn.occupation_a())
list_occ_beta = np.asarray(scf_wfn.occupation_b())


# Get orbital energies 
eps_a = np.asarray(scf_wfn.epsilon_a())
eps_b = np.asarray(scf_wfn.epsilon_b())
eps = np.append(eps_a, eps_b)


# Get orbital coefficients: 
Ca = np.asarray(scf_wfn.Ca())              
Cb = np.asarray(scf_wfn.Cb())              
C = np.block([ 
                [      Ca           ,   np.zeros_like(Cb) ],
                [np.zeros_like(Ca)  ,          Cb         ]
             ])


#-----------------------------------------------
#  Build Hamiltonian in second quantization:¶
#-----------------------------------------------

#-------- 0-body term: 
Hamiltonian_0body = E_nucl


#-------- 1-body term:
# Read from file:
#   H        =  np.array(scf_wfn.H())      
# or build core Hamiltonian
#
T = np.asarray(mints.ao_kinetic())
V = np.asarray(mints.ao_potential())
H_core_ao = T + V
Hamiltonian_1body  =  get_Hamiltonian_1body(H_core_ao,C)  


#--------- 2-body term
#   H_2body =  0.25* \sum
#Hamiltonian_2body = 0.25*gmo

# Get the two electron integrals using MintsHelper
Ints = np.array(mints.ao_eri())
gao  = get_gao(Ints)   #  chemists ->  physicists notation
gmo  = get_Hamiltonian_2body(gao,C)  ;  # C is over spin-orbitals 
Hamiltonian_2body = gmo 
#Hamiltonian_2body = get_Hamiltonian_2body(gao,get_spin_blocked(Ca))



#-----------------------------------------------
#  Now go ahead with FCI calcs  with two different Hamilaotnians:
#  1)  Full 2nd quantization hamiltonian
#  2)  Effective (frozen core) 2nd quantization  hamiltonian 
#
#-----------------------------------------------


#
# -- list of MOs
#--LiH
#

MSO_alpha_occ  = np.array([0   ])
MSO_alpha_virt = np.array([  1 ])

MSO_beta_occ   = np.array([2   ]) 
MSO_beta_virt  = np.array([  3 ])


MSO_occ =  np.append(MSO_alpha_occ,  MSO_beta_occ)
MSO_virt=  np.append(MSO_alpha_virt, MSO_beta_virt)

n_occ  = len(MSO_occ )
n_virt = len(MSO_virt)




#----------------------------------------------------------------
#  FCI energy
#
#----------------------------------------------------------------


from scipy.linalg  import eigh
from scipy.sparse import coo_matrix
import scipy.sparse.linalg 



help(get_FCI_hamiltonian)     

table_no_of_qbits         =  np.array([]) ;   # number of qubits / active MSO.
table_full_2nd_quant      =  np.array([]) ;   # save data from full 2nd quantization FCI
table_reduced_2nd_quant   =  np.array([]) ;   # save data from reduced (frozen core)  Hamitloanian
  

print("==================================================================================");
print("     12-qubits: Full Hamiltonian,  no-frozen core, all  vritual MSO included");
MSO_occ_tmp         = [0,   2  ] 
MSO_frozen_list_tmp = [        ]
MSO_active_list_tmp = [0, 1,2,3]

[FCI_list,fci_ham_sparse]=get_FCI_hamiltonian(
                          MSO_occ_tmp,
                          MSO_frozen_list_tmp,
                          MSO_active_list_tmp,
                          Hamiltonian_0body,Hamiltonian_1body,gmo) 

FCI_eigenvalues, FCI_eigenvectors= scipy.sparse.linalg.eigsh(fci_ham_sparse, k=4)



E_FCI =  FCI_eigenvalues[0]
E_FCI_corr = E_FCI - scf_e
print('Occupied MSO: ',MSO_occ_tmp)
print('Frozen   MSO: ',MSO_frozen_list_tmp)
print('Active   MSO: ',MSO_active_list_tmp)
print('E_HF           = ',scf_e)
print('E_FCI          = ',E_FCI)
print('E_FCI_corr     = ',E_FCI_corr)
print('FCI_list (basis) length             : ',len(FCI_list))
print('#of non-zeros in sparse hamiltonian : ',  fci_ham_sparse.getnnz())
print("  ");

#MSO_virt_tmp =  [ item for  item in  MSO_active_list_tmp  if item not in MSO_occ_tmp ]
#E_MP2_corr = get_MP2_corr(MSO_occ_tmp,MSO_virt_tmp,eps,gmo,False)
E_MP2_corr = get_MP2_corr(MSO_occ    ,MSO_virt    ,eps,gmo,False)
print('MP2 correlation energy: ', E_MP2_corr)
print('MP2 total energy:       ', scf_e +E_MP2_corr)


print("FCI_list = \n",FCI_list);
print("FCI_eigenvalues  =\n", FCI_eigenvalues) ;
print("FCI_eigenvectors =\n", FCI_eigenvectors) ;

table_no_of_qbits         =   np.append(table_no_of_qbits, len(MSO_active_list_tmp)) ; 
table_full_2nd_quant      =   np.append(table_full_2nd_quant,  E_FCI ) 
table_reduced_2nd_quant   =   np.append(table_reduced_2nd_quant, E_FCI) 

print("==================================================================================");


#-----------------------------------------------


print("\n\n\n");
print("    SUMMARY:  \n" );

print("No of quibits/active MSO        : ", table_no_of_qbits );
print("FCI energies, full       2nd Ham: ", table_full_2nd_quant ) ; 
print("FCI energies, fc reduced 2nd Ham: ", table_reduced_2nd_quant );


