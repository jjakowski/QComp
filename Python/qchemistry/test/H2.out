
*** tstart() called on mackaroo
*** at Fri Jul 31 17:25:36 2020

   => Loading Basis Set <=

    Name: STO-3G
    Role: ORBITAL
    Keyword: BASIS
    atoms 1-2 entry H          line    19 file /Users/j2c/anaconda/envs/OpenFermion-py37/share/psi4/basis/sto-3g.gbs 


         ---------------------------------------------------------
                                   SCF
               by Justin Turney, Rob Parrish, Andy Simmonett
                          and Daniel G. A. Smith
                              RHF Reference
                        1 Threads,    500 MiB Core
         ---------------------------------------------------------

  ==> Geometry <==

    Molecular point group: c1
    Full point group: D_inf_h

    Geometry (in Angstrom), charge = 0, multiplicity = 1:

       Center              X                  Y                   Z               Mass       
    ------------   -----------------  -----------------  -----------------  -----------------
         H            0.000000000000     0.000000000000    -0.373700000000     1.007825032230
         H            0.000000000000     0.000000000000     0.373700000000     1.007825032230

  Running in c1 symmetry.

  Rotational constants: A = ************  B =     59.88736  C =     59.88736 [cm^-1]
  Rotational constants: A = ************  B = 1795377.86645  C = 1795377.86645 [MHz]
  Nuclear repulsion =    0.708024097765588

  Charge       = 0
  Multiplicity = 1
  Electrons    = 2
  Nalpha       = 1
  Nbeta        = 1

  ==> Algorithm <==

  SCF Algorithm Type is PK.
  DIIS enabled.
  MOM disabled.
  Fractional occupation disabled.
  Guess Type is SAD.
  Energy threshold   = 1.00e-08
  Density threshold  = 1.00e-08
  Integral threshold = 0.00e+00

  ==> Primary Basis <==

  Basis Set: STO-3G
    Blend: STO-3G
    Number of shells: 2
    Number of basis function: 2
    Number of Cartesian functions: 2
    Spherical Harmonics?: true
    Max angular momentum: 0

  ==> Pre-Iterations <==

   -------------------------------------------------------
    Irrep   Nso     Nmo     Nalpha   Nbeta   Ndocc  Nsocc
   -------------------------------------------------------
     A          2       2       0       0       0       0
   -------------------------------------------------------
    Total       2       2       1       1       1       0
   -------------------------------------------------------

  ==> Integral Setup <==

  Using in-core PK algorithm.
   Calculation information:
      Number of atoms:                   2
      Number of AO shells:               2
      Number of primitives:              6
      Number of atomic orbitals:         2
      Number of basis functions:         2

      Integral cutoff                 1.00e-12
      Number of threads:                 1

  Performing in-core PK
  Using 12 doubles for integral storage.
  We computed 6 shell quartets total.
  Whereas there are 6 unique shell quartets.

  ==> DiskJK: Disk-Based J/K Matrices <==

    J tasked:                  Yes
    K tasked:                  Yes
    wK tasked:                  No
    Memory [MiB]:              375
    Schwarz Cutoff:          1E-12

    OpenMP threads:              1

  Minimum eigenvalue in the overlap matrix is 3.4496680357E-01.
  Using Symmetric Orthogonalization.

  SCF Guess: Superposition of Atomic Densities via on-the-fly atomic UHF.

  ==> Iterations <==

                        Total Energy        Delta E     RMS |[F,P]|

   @RHF iter SAD:    -0.71616892688555   -7.16169e-01   0.00000e+00 
   @RHF iter   1:    -1.11632556446324   -4.00157e-01   0.00000e+00 DIIS
   @RHF iter   2:    -1.11632556446324    0.00000e+00   0.00000e+00 DIIS
  Energy and wave function converged.


  ==> Post-Iterations <==

    Orbital Energies [Eh]
    ---------------------

    Doubly Occupied:                                                      

       1A     -0.575503  

    Virtual:                                                              

       2A      0.663551  

    Final Occupation by Irrep:
              A 
    DOCC [     1 ]

  @RHF Final Energy:    -1.11632556446324

   => Energetics <=

    Nuclear Repulsion Energy =              0.7080240977655875
    One-Electron Energy =                  -2.4976936071679474
    Two-Electron Energy =                   0.6733439449391169
    Total Energy =                         -1.1163255644632430

Computation Completed


Properties will be evaluated at   0.000000,   0.000000,   0.000000 [a0]

Properties computed using the SCF density matrix

  Nuclear Dipole Moment: [e a0]
     X:     0.0000      Y:     0.0000      Z:     0.0000

  Electronic Dipole Moment: [e a0]
     X:     0.0000      Y:     0.0000      Z:     0.0000

  Dipole Moment: [e a0]
     X:     0.0000      Y:     0.0000      Z:     0.0000     Total:     0.0000

  Dipole Moment: [D]
     X:     0.0000      Y:     0.0000      Z:     0.0000     Total:     0.0000


*** tstop() called on mackaroo at Fri Jul 31 17:25:36 2020
Module time:
	user time   =       0.17 seconds =       0.00 minutes
	system time =       0.01 seconds =       0.00 minutes
	total time  =          0 seconds =       0.00 minutes
Total time:
	user time   =       0.17 seconds =       0.00 minutes
	system time =       0.01 seconds =       0.00 minutes
	total time  =          0 seconds =       0.00 minutes
E(SCF): -1.116325564463243 <psi4.core.RHF object at 0x1a2d6b44d0>

Alpha orbital energies:  [-0.57550286  0.66355144]
Beta  orbital energies:  [-0.57550286  0.66355144]

N-alpha-electrons:  1 
N-beta--electrons:  1 

Alpha occupations:  [1. 0.]
Beta  occupations:  [1. 0.]

Nuclear repulsion: 0.7080240977655875
Help on function get_FCI_hamiltonian in module qchemistry.qcpy:

get_FCI_hamiltonian(MSO_occ, MSO_frozen_list, MSO_active_list, Hamiltonian_0body, Hamiltonian_1body, gmo)
    input:  MSO_occ, MSO_frozen_list, MSO_active_list, 
           Hamiltonian_0body, Hamiltonian_1body, gmo
           where gmo(p,q,r,s) =  <p,q||r,s>     
    output: FCI_list, fci_ham_sparse

==================================================================================
     12-qubits: Full Hamiltonian,  no-frozen core, all  vritual MSO included
Occupied MSO:  [0, 2]
Frozen   MSO:  []
Active   MSO:  [0, 1, 2, 3]
E_HF           =  -1.116325564463243
E_FCI          =  -1.1371757102334599
E_FCI_corr     =  -0.02085014577021682
FCI_list (basis) length             :  6
#of non-zeros in sparse hamiltonian :  10
  
MP2 correlation energy:  -0.013311668818382225
MP2 total energy:        -1.1296372332816254
FCI_list = 
 [[0, 1], [0, 2], [0, 3], [1, 2], [1, 3], [2, 3]]
FCI_eigenvalues  =
 [-1.13717571 -0.53969856 -0.53969856 -0.53969856]
FCI_eigenvectors =
 [[-2.70755030e-17  8.10024284e-01  3.11514870e-02  5.85568309e-01]
 [ 9.93475180e-01 -3.20392668e-17  0.00000000e+00 -3.51764307e-18]
 [ 4.82643830e-17 -6.80596869e-03  7.06512622e-01 -2.81707849e-02]
 [-3.16134206e-17  6.80596869e-03 -7.06512622e-01  2.81707849e-02]
 [-1.14048531e-01 -3.25899695e-16  0.00000000e+00 -2.87337526e-17]
 [ 9.73580105e-18 -5.86317335e-01  2.66348178e-02  8.09643483e-01]]
==================================================================================




    SUMMARY:  

No of quibits/active MSO        :  [4.]
FCI energies, full       2nd Ham:  [-1.13717571]
FCI energies, fc reduced 2nd Ham:  [-1.13717571]
